wave_kinematics/src/quadrotor.cpp

```diff
@@ -0,0 +1,435 @@
+    this->states(4) = 0.0;      // angular velocity y
+    this->states(5) = 0.0;      // angular velocity z
+    this->states(6) = pose(0);  // x
+    this->states(7) = pose(1);  // y
+    this->states(8) = pose(2);  // z
+    this->states(9) = 0.0;      // linear velocity x
+    this->states(10) = 0.0;     // linear velocity y
+    this->states(11) = 0.0;     // linear velocity z
+
+    this->Ix = 0.0963;  // inertial x
+    this->Iy = 0.0963;  // inertial y
+    this->Iz = 0.1927;  // inertial z
+
+    this->kr = 0.1;  // rotation drag constant
+    this->kt = 0.2;  // translation drag constant;
+
+    this->l = 0.9;  // arm length
+    this->d = 1.0;  // drag
+
+    this->m = 1.0;   // mass of quad
+    this->g = 10.0;  // gravitational constant
+
+    this->attitude_setpoints = Vec4();
+    this->attitude_setpoints << 0.0, 0.0, 0.0, 0.5;
+
+    this->position_setpoints = VecX(3);
+    this->position_setpoints << pose(0), pose(1), pose(2);
+
+    this->attitude_controller = AttitudeController();
+    this->position_controller = PositionController();
+}
+
+int QuadrotorModel::update(VecX motor_inputs, double dt) {
+    double ph, th, ps;
+    double p, q, r;
+    double x, y, z;
+    double vx, vy, vz;
+
+    double Ix, Iy, Iz;
+    double kr, kt;
+    double tauf, taup, tauq, taur;
+    double m, g;
+
+    Mat4 A;
+    Vec4 tau;
+
+    // setup
+    ph = this->states(0);
+    th = this->states(1);
+    ps = this->states(2);
+
+    p = this->states(3);
+    q = this->states(4);
+    r = this->states(5);
+
+    x = this->states(6);
+    y = this->states(7);
+    z = this->states(8);
+
+    vx = this->states(9);
+    vy = this->states(10);
+    vz = this->states(11);
+
+    Ix = this->Ix;
+    Iy = this->Iy;
+    Iz = this->Iz;
+
+    kr = this->kr;
+    kt = this->kt;
+
+    m = this->m;
+    g = this->g;
+
+    // convert motor inputs to angular p, q, r and total thrust
+    // clang-format off
+  A << 1.0, 1.0, 1.0, 1.0,
+       0.0, -this->l, 0.0, this->l,
+       -this->l, 0.0, this->l, 0.0,
+       -this->d, this->d, -this->d, this->d;
+    // clang-format on
+
+    tau = A * motor_inputs;
+    tauf = tau(0);
+    taup = tau(1);
+    tauq = tau(2);
+    taur = tau(3);
+
+    // update
+    // clang-format off
+  this->states(0) = ph + (p + q * sin(ph) * tan(th) + r * cos(ph) * tan(th)) * dt;
+  this->states(1) = th + (q * cos(ph) - r * sin(ph)) * dt;
+  this->states(2) = ps + ((1 / cos(th)) * (q * sin(ph) + r * cos(ph))) * dt;
+  this->states(3) = p + (-((Iz - Iy) / Ix) * q * r - (kr * p / Ix) + (1 / Ix) * taup) * dt;
+  this->states(4) = q + (-((Ix - Iz) / Iy) * p * r - (kr * q / Iy) + (1 / Iy) * tauq) * dt;
+  this->states(5) = r + (-((Iy - Ix) / Iz) * p * q - (kr * r / Iz) + (1 / Iz) * taur) * dt;
+  this->states(6) = x + vx * dt;
+  this->states(7) = y + vy * dt;
+  this->states(8) = z + vz * dt;
+  this->states(9) = vx + ((-kt * vx / m) + (1 / m) * (cos(ph) * sin(th) * cos(ps) + sin(ph) * sin(ps)) * tauf) * dt;
```

Comment:
Mention what equation you're using?
Seems like it could be simplified by having 3d vectors for velocity and position, etc. and operating on those with a rotation matrix (assuming that's what is happening here)